View Compilation build error in VS15.7 due to missing references

I have a solution with a large number of projects that I am porting to MvcCore. In this solution I have a number of class libraries that contains views. I was using RazorGenerator.MsBuild in the MVC5 days and some view engine hackery to get the main web project to do what you are enabling in 2.1.

I've been trying out the view precompilation using the 2.1.0-preview1, and I'm using VS 15.7 preview 2.

I got the code to the point where it compiles. From the command line invoking dotnet build and dotnet run I haven't had any issues, it is working well.

The issue I am having is building within Visual Studio I am getting build failures on incremental builds due to missing project references being added to the csc command line. This may not strictly be a View Compilation bug, but I am not sure what else to check.

I've installed the Project System extension and looked at the build log. It appears that the ResolveProjectReferences isn't being called to populate these dll's and then has the build failure.

This only seems to be affecting projects containing views where I have enabled the RazorSdk and and RazorCompileOnBuild=true settings. The csc command does appear to be including PackageReferences.

Strangely according to the Build Logging it is when building the main project that has references to the class libraries, that it is failing.

I've managed to reproduce this in a clean class library project. What I did was create a new project on the 2.0 templates, upgraded to 2.1 and created a class library for the views. I've created a common class library and moved the Web models to the Common library. Then I've moved the Account views into the View library. This is referenced from the web project. It seemed to work OK for a single project, however when duplicating the class library I get the same build error in my full solution.

Sometimes it is a bit hard to trigger the bug. The build output looks like this
```
Error occurred while restoring NuGet packages: Object reference not set to an instance of an object.
1>------ Build started: Project: ViewClass, Configuration: Debug Any CPU ------
1>ViewClass -> C:\dev\BugReport\ViewClass\bin\Debug\netstandard2.0\ViewClass.dll
1>ViewClass -> C:\dev\BugReport\ViewClass\bin\Debug\netstandard2.0\ViewClass.PrecompiledViews.dll
2>------ Build started: Project: BugReport, Configuration: Debug Any CPU ------
2>Class1.cs(1,7,1,17): error CS0246: The type or namespace name 'CommonCode' could not be found (are you missing a using directive or an assembly reference?)
2>Class1.cs(12,30,12,39): error CS0246: The type or namespace name 'SomeClass' could not be found (are you missing a using directive or an assembly reference?)
2>Class1.cs(14,30,14,34): error CS0115: 'SomeOther.Test()': no suitable method found to override
2>Done building project "ViewClassSecond.csproj" -- FAILED.
2>Class1.cs(1,7,1,17): error CS0246: The type or namespace name 'CommonCode' could not be found (are you missing a using directive or an assembly reference?)
2>Class1.cs(12,30,12,39): error CS0246: The type or namespace name 'SomeClass' could not be found (are you missing a using directive or an assembly reference?)
2>Class1.cs(14,30,14,34): error CS0115: 'SomeOther.Test()': no suitable method found to override
2>Done building project "ViewClassThird.csproj" -- FAILED.
========== Build: 1 succeeded, 1 failed, 3 up-to-date, 0 skipped ==========
```
I've attached the binlog and this is the screenshot. It appears that the project is pulling in the dependency, trying to build the class libraries and this is the scenario it is failing.

![image](https://user-images.githubusercontent.com/2775804/38476352-a39f899a-3bf0-11e8-91d0-ae16331504d4.png)

[BugReport_Debug_AnyCPU_Build_2018-04-09T12_20_01.9597168+10_00.binlog.zip](https://github.com/aspnet/Razor/files/1888268/BugReport_Debug_AnyCPU_Build_2018-04-09T12_20_01.9597168.10_00.binlog.zip)

I'll upload the solution that I managed to reproduce to a new git repo.

--- a/src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets
+++ b/src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets
@@ -69,6 +69,11 @@ Copyright (c) .NET Foundation. All rights reserved.
       _RazorAddDebugSymbolsProjectOutputGroupOutput
     </DebugSymbolsProjectOutputGroupDependsOn>
 
+    <PrepareForRunDependsOn>
+      _RazorPrepareForRun;
+      $(PrepareForRunDependsOn)
+    </PrepareForRunDependsOn>
+
   </PropertyGroup>
 
   <!-- 
@@ -362,13 +367,21 @@ Copyright (c) .NET Foundation. All rights reserved.
   </Target>
 
   <!--
+    Set up RazorCompile to run before PrepareForRun. This should ensure that the Razor dll and pdbs are available to be copied
+    as part of GetCopyToOutputDirectoryItems which is invoked during PrepareForRun.
+  -->
+  <Target
+    Name="_RazorPrepareForRun"
+    DependsOnTargets="RazorCompile"
+    Condition="'$(ResolvedRazorCompileToolset)'=='RazorSdk' and '$(RazorCompileOnBuild)'=='true'" />
+
+  <!--
     Called as part of GetCopyToOutputDirectoryItems - this target populates the list of items that get
     copied to the output directory when building as a project reference.
   -->
   <Target 
     Name="_RazorGetCopyToOutputDirectoryItems"
     BeforeTargets="GetCopyToOutputDirectoryItems"
-    DependsOnTargets="RazorCompile"
     Condition="'$(ResolvedRazorCompileToolset)'=='RazorSdk' and '$(RazorCompileOnBuild)'=='true'">
 
     <!-- 

--- a/test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs
+++ b/test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs
@@ -1,6 +1,7 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.IO;
 using System.Linq;
 using System.Reflection;
@@ -470,6 +471,41 @@ namespace Microsoft.AspNetCore.Razor.Design.IntegrationTests
 
             Assert.FileDoesNotExist(result, OutputPath, "SimpleMvcFSharp.Views.dll");
             Assert.FileDoesNotExist(result, OutputPath, "SimpleMvcFSharp.Views.pdb");
+        }
+
+        [Fact]
+        [InitializeTestProject("AppWithP2PReference", additionalProjects: new[] { "ClassLibrary", "ClassLibrary2" })]
+        public async Task Build_WithP2P_WorksWhenBuildProjectReferencesIsDisabled()
+        {
+            // Simulates building the same way VS does by setting BuildProjectReferences=false.
+            // With this flag, P2P references aren't resolved during GetCopyToOutputDirectoryItems. This test ensures that
+            // no Razor work is done in such a scenario and the build succeeds.
+            var additionalProjectContent = @"
+<ItemGroup>
+  <ProjectReference Include=""..\ClassLibrary2\ClassLibrary2.csproj"" />
+</ItemGroup>
+";
+            AddProjectFileContent(additionalProjectContent);
+
+            var result = await DotnetMSBuild(target: default);
+
+            Assert.BuildPassed(result);
+
+            Assert.FileExists(result, OutputPath, "AppWithP2PReference.dll");
+            Assert.FileExists(result, OutputPath, "AppWithP2PReference.Views.dll");
+            Assert.FileExists(result, OutputPath, "ClassLibrary.dll");
+            Assert.FileExists(result, OutputPath, "ClassLibrary.Views.dll");
+            Assert.FileExists(result, OutputPath, "ClassLibrary2.dll");
+            Assert.FileExists(result, OutputPath, "ClassLibrary2.Views.dll");
+
+            // Force a rebuild of ClassLibrary2 by changing a file
+            var class2Path = Path.Combine(Project.SolutionPath, "ClassLibrary2", "Class2.cs");
+            File.AppendAllText(class2Path, Environment.NewLine + "// Some changes");
+
+            // dotnet msbuild /p:BuildProjectReferences=false
+            result = await DotnetMSBuild(target: default, "/p:BuildProjectReferences=false", suppressRestore: true);
+
+            Assert.BuildPassed(result);
         }
 
         private static DependencyContext ReadDependencyContext(string depsFilePath)

--- a/test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/InitializeTestProjectAttribute.cs
+++ b/test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/InitializeTestProjectAttribute.cs
@@ -37,7 +37,7 @@ namespace Microsoft.AspNetCore.Razor.Design.IntegrationTests
             }
 
             MSBuildIntegrationTestBase.Project = ProjectDirectory.Create(_originalProjectName, _testProjectName, _baseDirectory, _additionalProjects, _language);
-            MSBuildIntegrationTestBase.TargetFramework = _originalProjectName == "ClassLibrary" ? "netstandard2.0" : "netcoreapp2.0";
+            MSBuildIntegrationTestBase.TargetFramework = _originalProjectName.StartsWith("ClassLibrary") ? "netstandard2.0" : "netcoreapp2.0";
         }
 
         public override void After(MethodInfo methodUnderTest)

--- a/test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/MSBuildIntegrationTestBase.cs
+++ b/test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/MSBuildIntegrationTestBase.cs
@@ -79,7 +79,12 @@ namespace Microsoft.AspNetCore.Razor.Design.IntegrationTests
                 buildArgumentList.Add($"/p:_RazorSuppressCurrentUserOnlyPipeOptions=true");
             }
 
-            buildArgumentList.Add($"/t:{target} /p:Configuration={Configuration} {args}");
+            if (!string.IsNullOrEmpty(target))
+            {
+                buildArgumentList.Add($"/t:{target}");
+            }
+
+            buildArgumentList.Add($"/p:Configuration={Configuration} {args}");
             var buildArguments = string.Join(" ", buildArgumentList);
 
             return MSBuildProcessManager.RunProcessAsync(

--- a/test/testapps/ClassLibrary2/Class2.cs
+++ b/test/testapps/ClassLibrary2/Class2.cs
@@ -1,0 +1,13 @@
+using System;
+using ClassLibrary;
+
+namespace ClassLibrary2
+{
+    public class Class2
+    {
+        public void Method()
+        {
+            Console.WriteLine(typeof(Class1));
+        }
+    }
+}

--- a/test/testapps/ClassLibrary2/ClassLibrary2.csproj
+++ b/test/testapps/ClassLibrary2/ClassLibrary2.csproj
@@ -1,0 +1,27 @@
+<Project Sdk="Microsoft.NET.Sdk.Razor">
+  
+  <PropertyGroup>
+    <_RazorMSBuildRoot>$(SolutionRoot)src\Microsoft.AspNetCore.Razor.Design\bin\$(Configuration)\netstandard2.0\</_RazorMSBuildRoot>
+  </PropertyGroup>
+
+  <Import Project="$(SolutionRoot)src\Microsoft.AspNetCore.Razor.Design\build\netstandard2.0\Microsoft.AspNetCore.Razor.Design.props" />
+
+  <PropertyGroup>
+    <!-- Override for the MVC extension -->
+    <_MvcExtensionAssemblyPath>$(SolutionRoot)src\Microsoft.AspNetCore.Mvc.Razor.Extensions\bin\$(Configuration)\netstandard2.0\Microsoft.AspNetCore.Mvc.Razor.Extensions.dll</_MvcExtensionAssemblyPath>
+  </PropertyGroup>
+  <Import Project="$(SolutionRoot)src\Microsoft.AspNetCore.Mvc.Razor.Extensions\build\netstandard2.0\Microsoft.AspNetCore.Mvc.Razor.Extensions.props" />
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+  
+  <ItemGroup>
+    <ProjectReference Include="..\ClassLibrary\ClassLibrary.csproj" />
+  </ItemGroup>
+
+  <!-- Test Placeholder -->
+
+  <Import Project="$(SolutionRoot)src\Microsoft.AspNetCore.Mvc.Razor.Extensions\build\netstandard2.0\Microsoft.AspNetCore.Mvc.Razor.Extensions.targets" />
+
+</Project>

--- a/test/testapps/ClassLibrary2/Views/Shared/Index.cshtml
+++ b/test/testapps/ClassLibrary2/Views/Shared/Index.cshtml
@@ -1,0 +1,2 @@
+﻿@{ var message = "Hello world";}
+@message

--- a/test/testapps/ClassLibrary2/Views/_ViewImports.cshtml
+++ b/test/testapps/ClassLibrary2/Views/_ViewImports.cshtml
@@ -1,0 +1,2 @@
+@using ClassLibrary
+@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers

